### ISSUE TITLE
pathlib.Path.open() pyright different return type from mypy

### ISSUE THREAD

**Issue by timkgh** [microsoft/pyright#460]
Mac, python 3.7.6
`pathlib.Path.open(mode='wt', encoding='utf-8') -> TextIO` in `mypy` but `-> IO[Any]` in `pyright`
Feels like TextIO is the correct one here, but not sure if I don't have pyright configured correctly or it's a bug.

**Comment by timkgh** [comment]
I should mention this is when used as a context:
```
p = pathlib.Path('/tmp/a.txt')
with p.open(mode='wt', encoding='utf-8') as f:
    reveal_type(f)
```
Maybe it's the same issue as #459?

**Comment by erictraut** [comment]
The type stub that ships with Pyright is clear that the return type is IO[Any] in this case:

```
    def open(self, mode: str = ..., buffering: int = ...,
             encoding: Optional[str] = ..., errors: Optional[str] = ...,
             newline: Optional[str] = ...) -> IO[Any]: ...
```

It's possible that the typeshed type stub for `pathlib` has been updated recently. I haven't synched with the typeshed repo in over a month. I'll incorporate the latest changes. That will likely address this issue. In the meantime, you can download your own copy of the typeshed stubs and configure Pyright to use them.

**Comment by timkgh** [comment]
Just looked and the latest version in typeshed still shows it as IO[Any]
https://github.com/python/typeshed/blob/master/stdlib/3/pathlib.pyi#L94
I also looked at the typeshed that mypy comes with and it's the same IO[Any] but
somehow mypy seems to understand the open mode, for `wb` the type is `BinaryIO`.

**Comment by erictraut** [comment]
I've confirmed that mypy includes some special-case logic that overrides the type stub in this case. I've avoided adding any such special-case logic in Pyright, and I'm very reluctant to do so now. I'd rather fix the type stubs. Until recently, it wouldn't have been possible to provide the correct type definitions in the type stub, but now that Literal has been introduced, it is possible.

If this is important to you, I recommend that you submit a PR to the typeshed repo that introduces overloaded versions of Path.open that return TextIO or BinaryIO based on the specified mode string literal. Once your PR is accepted to the typeshed repo, let me know, and I'll pull it into Pyright.
